### PR TITLE
faster syncing

### PR DESCRIPTION
**Pull request by tersec**

### CODE REVIEW

**Comment by tersec** [comment]
Extacting `sync_speed` numbers, where `devel` is status quo and `nfa` is this PR.
```
$ <devel_branch.txt datamash mean 1 q1 1 q3 1 sstdev 1 count 1
1.905	1.5442	2.0832	0.93201618866707	65
$ <nfa_branch.txt datamash mean 1 q1 1 q3 1 sstdev 1 count 1
4.5051620437956	3.4879	6.7727	2.1925928981302	137
```

File: beacon_chain/attestation_pool.nim [hunk 1]
```diff
@@ -227,24 +220,25 @@ proc addResolved(pool: var AttestationPool, blck: BlockRef, attestation: Attesta
...
+  if not isValidAttestationTargetEpoch(
+      attestation.data.slot.compute_epoch_at_slot, attestation.data):
     notice "Invalid attestation",
       attestation = shortLog(attestation),
-      current_epoch = get_current_epoch(state)
+      current_epoch = attestation.data.slot.compute_epoch_at_slot
     return
 
+  # Get a temporary state at the (block, slot) targeted by the attestation
+  debugEcho "FOO1: ", blck.root, "; ", attestation.data.slot
```
Comment (arnetheduck) [inline beacon_chain/attestation_pool.nim:235]: ...

File: beacon_chain/attestation_pool.nim [hunk 2]
```diff
@@ -227,24 +220,25 @@ proc addResolved(pool: var AttestationPool, blck: BlockRef, attestation: Attesta
...
-  template state(): BeaconState = pool.blockPool.tmpState.data.data
-
   # Check that the attestation is indeed valid
   # TODO: we might want to split checks that depend
   #       on the state and those that don't to cheaply
   #       discard invalid attestations before rewinding state.
-
-  if not isValidAttestationTargetEpoch(state, attestation.data):
+  if not isValidAttestationTargetEpoch(
+      attestation.data.slot.compute_epoch_at_slot, attestation.data):
```
Comment (arnetheduck) [inline beacon_chain/attestation_pool.nim:228]: this is a bug actually that has been around for a while: we should be using `attestation.data.target.epoch` here which incidentally will collapse the value correctly to a single epoch

Comment (tersec) [reply beacon_chain/attestation_pool.nim]: https://github.com/status-im/nim-beacon-chain/pull/1348/commits/eabda114e834d476d908acc019301d3d66aa1151

File: beacon_chain/attestation_pool.nim (hunk 1 above)
Comment (tersec) [reply beacon_chain/attestation_pool.nim]: https://github.com/status-im/nim-beacon-chain/pull/1348/commits/eabda114e834d476d908acc019301d3d66aa1151

File: beacon_chain/block_pools/candidate_chains.nim [hunk 3]
```diff
@@ -30,7 +30,8 @@ proc putBlock*(
   dag.db.putBlock(signedBlock)
 
 proc updateStateData*(
-  dag: CandidateChains, state: var StateData, bs: BlockSlot) {.gcsafe.}
+  dag: CandidateChains, state: var StateData, bs: BlockSlot,
+  matchEpoch: bool = false) {.gcsafe.}
```
Comment (arnetheduck) [inline beacon_chain/block_pools/candidate_chains.nim:34]: where is the call with matchEpoch = true?

Comment (tersec) [reply beacon_chain/block_pools/candidate_chains.nim:34]: https://github.com/status-im/nim-beacon-chain/pull/1348/commits/06c8b1dd329c2344f4b08ef83475506e7270f1ee

File: beacon_chain/block_pools/candidate_chains.nim [hunk 4]
```diff
@@ -360,42 +361,26 @@ proc getState(
...
-func getStateCacheIndex(dag: CandidateChains, blockRoot: Eth2Digest, slot: Slot): int =
+func getStateCacheIndex(
+    dag: CandidateChains, blockRoot: Eth2Digest, slot: Slot, matchEpoch: bool):
+    int =
   for i, cachedState in dag.cachedStates:
     let (cacheBlockRoot, cacheSlot, _) = cachedState
-    if cacheBlockRoot == blockRoot and cacheSlot == slot:
+    if cacheBlockRoot == blockRoot and (cacheSlot == slot or
+        (matchEpoch and
+          cacheSlot.compute_epoch_at_slot == slot.compute_epoch_at_slot)):
```
Comment (arnetheduck) [inline beacon_chain/block_pools/candidate_chains.nim:371]: technically, if we're matching epoch, any ancestor block within the same history as the given block root would be a match

Comment (tersec) [reply beacon_chain/block_pools/candidate_chains.nim:371]: Within `MAX_SEED_LOOKAHEAD` epochs or so, sure.

**Pull request merged**